feat(rest): allow basePath for rest servers

raymondfeng opened the pull request:
See https://github.com/strongloop/loopback-next/issues/918

Some deviation from #918:

- The `basePath` is only available as part of `RestServerConfig` because we set up express routes within the constructor.
- The `basePath` only applies to registered routes (including static assets)
- `servers` in openapi spec are adjusted accordingly
 
<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

bajtos commented on packages/rest/test/integration/rest.server.integration.ts:
```diff
@@ -701,6 +701,39 @@ paths:
+    const response = await createClientForHandler(server.requestHandler).get(
+      '/openapi.json',
+    );
+    expect(response.body.servers).to.containEql({url: '/api'});
```
Single logical assert per test please.

For example:

```ts
describe('basePath', () => {
  it('applies basePath to controller routes', () => {
    // ...
  });

  it('applies basePath to static assets', () => {
    // ...
  });

  it('applies basePath to OpenAPI spec "servers"', () => {
    // ...
  });
});
```

bajtos commented:
> The basePath is only available as part of RestServerConfig because we set up express routes within the constructor.

Can this limitation affect usability of the feature when configuring basePath from an application?

Please add an acceptance test showing how a RestApplication can configure basePath. Try to mimic the coding conventions we are using in our example applications (e.g. [examples/todo/src/application.ts](https://github.com/strongloop/loopback-next/blob/ad0229b6711184da5a00c66ddeba585afe89eb79/examples/todo/src/application.ts)) - a custom class extending RestApplication where the configuration options are modified inside application's constructor.

bajtos commented:
> The basePath is only available as part of RestServerConfig because we set up express routes within the constructor.

Also: have you considered to refactor RestServer so that setup of express routes is deferred until we need to access the request-handler function?

raymondfeng commented:
> Can this limitation affect usability of the feature when configuring basePath from an application?

I don't think so. IMO, `basePath` is in the same category as `host` and `port` for REST servers and we don't have APIs to set up such values beyond configuration.

Anyway, I refactored `rest-server.ts` a bit so that the Express app will only be set up if needed. I also added `basePath()` to RestApplication and RestServer.

PTAL.

raymondfeng commented on packages/rest/src/rest.server.ts:
```diff
@@ -741,13 +760,31 @@ export class RestServer extends Context implements Server, HttpServerLike {
+   */
+  basePath(path: string = '') {
+    if (this._requestHandler) {
+      throw new Error(
```
Making LoopBack 4 core reactive to dynamic artifact registration is a much bigger fish to hunt. I have some ideas to achieve that outside of the scope of this PR. Let's defer it to that discussion.

bajtos replied:
Fair enough 👍

raymondfeng commented on packages/rest/src/rest.server.ts:
```diff
@@ -374,6 +385,14 @@ export class RestServer extends Context implements Server, HttpServerLike {
+    if (specObj.servers && this._basePath) {
+      for (const s of specObj.servers) {
+        if (s.url.startsWith('/')) {
+          s.url = s.url === '/' ? this._basePath : this._basePath + s.url;
```
It's not practical to infer `basePath` from `servers`. Please note the `spec.servers` lists `public` urls of REST APIs, which can be very different from the `basePath` of a LoopBack 4 RestServer instance.

We should use `basePath` to drive the `default` server urls, not the other way around. I agree that we can just fix `/`.

bajtos commented on packages/rest/test/integration/rest.application.integration.ts:
```diff
@@ -92,6 +92,15 @@ describe('RestApplication (integration)', () => {
+    restApp.static('/', ASSETS);
+    await restApp.start();
+    client = createRestAppClient(restApp);
+    await client.get('/html/index.html').expect(200);
```
I find this test very confusing. As far as I understand `givenApplication()` helper, it creates an app with no API routes configured. As I understand `app.basePath`, it's intended primarily for customizing the base path of REST APIs.

I see that there are other tests verifying the impact of basePath on different kinds of routes, I agree there is no need to duplicate those tests in `rest.application.integration.ts`.

Can you please add a comment here? Explain that in this test, we are invoking a route for static assets, because basePath applies to all kinds of routes, including static assets.

Alternatively, rework this test to use a handler-function route (see the example below) - I think it will make the intent of this test much easier to understand even without comments.

```ts
app.route('get', '/status', {/*spec*/}, () => {running: true});
// ...
client.get('/api/status').expect(200, {running: true});
```

bajtos commented on packages/rest/test/unit/rest.server/rest.server.unit.ts:
```diff
@@ -86,6 +86,40 @@ describe('RestServer', () => {
+      app.component(RestComponent);
+      const server = await app.getServer(RestServer);
+      expect(() => {
+        if (server.requestHandler) {
```
If this condition happens to be false, then `server.basePath` is not invoked at all and the test fails with a confusing error.

Please convert this `if` into an explicit assertion. For example:

```ts
expect.assert(!!server.requestHandler, 'requestHandler should have been set up by now');
expect(() => server.basePath('/api')).to.throw(/*...*/);
```

AnanthGopal commented:
hi @raymondfeng 

Thanks to every loopback developer. when we will use this "**Add basePath configuration**"  feature ?
